### PR TITLE
Fix from profiling

### PR DESCRIPTION
**Pull request by Cmurilochem**
This PR attempts to make some fixes to improve performance of the hand written implementation of the Poisson problem based on profiling.

Changes made:
- [x] Add comments to hand written Poisson example.
- [x] Fix in some type-instabilities associated with `node_coordinates_from_monomials_exponents` and ` monomial_exponents_from_filter` (mesh.jl)
- [x] Add caches to sub-types of `AbstractSpace`: `IsoParametricSpace` and `LagrangeSpace`
- [x] Refactor `same_valid_ids` from mesh.jl

### CODE REVIEW

File: src/interpolation.jl [hunk 1]
```diff
@@ -1620,32 +1651,58 @@ function lagrange_space(domain,order;
...
+function setup_space(space::AbstractSpace)
+    if space.cache !== nothing
+        return space
+    end
+    face_dofs = GT.face_dofs(space)
+    cache = space_cache(;face_dofs)
+    replace_cache(space,cache)
+end
 
+function replace_cache(space::AbstractSpace,cache)
```
Comment (fverdugo) [inline src/interpolation.jl:1668]: Hi @Cmurilochem This should be only for LagrangeSpace

File: src/mesh.jl [hunk 2]
```diff
@@ -1387,7 +1384,7 @@ function fill_face_boundary_mesh_topology!(topo,mesh,D,d)
                     dfaces = vertex_to_dfaces[vertex]
                     for dface1 in dfaces
                         vertices1 = dface_to_vertices[dface1]
-                        if same_valid_ids(vertices,vertices1)
+                        if same_valid_ids(collect(vertices),collect(vertices1))
```
Comment (fverdugo) [inline src/mesh.jl:1387]: By adding collect, you are allocating a new array (we want to avoid this). The same is also true for other instances below.

Comment (Cmurilochem) [reply src/mesh.jl]: Thanks; very useful info: I removed recollect but now the vector types must be `AbstractVector`s I suppose.

File: src/interpolation.jl (hunk 1 above)
Comment (Cmurilochem) [reply src/interpolation.jl]: Thanks @fverdugo for pointing this out; that is true indeed! Fixed!

File: GalerkinToolkitExamples/src/poisson.jl [hunk 3]
```diff
@@ -263,90 +298,120 @@ function main_hand_written(params)
     Tgrad = SVector{D,T}
     Tx = Tgrad
     TJ = SMatrix{D,D,T}
```
Comment (fverdugo) [inline GalerkinToolkitExamples/src/poisson.jl:300]: Replace `SMatrix{D,D,T}` with `typeof(zero(Tx)*transpose(zero(Tx)))`. The first one is not a concrete type (my bad).

File: src/mesh.jl [hunk 4]
```diff
@@ -1521,31 +1519,36 @@ function intersection!(a,b,na,nb)
   end
 end
 
-function same_valid_ids(a,b)
-  function is_subset(a,b)
+function find_eq(v::T, b::AbstractVector{T}) where T
```
Comment (fverdugo) [inline src/mesh.jl:1522]: I am not sure about these changes. Why do you need them?

File: GalerkinToolkitExamples/src/poisson.jl (hunk 3 above)
Comment (Cmurilochem) [reply GalerkinToolkitExamples/src/poisson.jl:300]: Thanks! done!

File: src/mesh.jl (hunk 4 above)
Comment (Cmurilochem) [reply src/mesh.jl:1522]: You are right. I removed the type declarations. The instability here was mainly due to the operations on the `c` quantity in different scope and also due to nested functions.

**Comment by Cmurilochem** [comment]
@fverdugo,

I am working now on resolving other type-instabilities, *e.g.*, from `generate_dof_ids_step_1(...)` in interpolation.jl. But I want do it in another PR. Please, feel free to suggest any further changes of the things I've done so far.

I am excited to see if we managed to get a better performance in the benchmarks with the caches.

Thanks for your help.

**Review by fverdugo — APPROVED** [review]

**Pull request merged**